types/src/v0/impls/header.rs

```diff
@@ -132,6 +77,276 @@ impl Committable for Header {
+                    EitherOrVersion::Right(commit) => Ok(Header::V1(
+                        v0_1::Header::deserialize_with_chain_config(commit.into(), seq)?,
+                    )),
+                    // For all versions > 0.1, the first "field" is not actually part of the `Header` struct.
+                    // We just delegate directly to the derived deserialization impl for the appropriate version.
+                    EitherOrVersion::Version(Version { major: 0, minor: 2 }) => Ok(Header::V2(
+                        seq.next_element()?
+                            .ok_or_else(|| de::Error::missing_field("fields"))?,
+                    )),
+                    EitherOrVersion::Version(Version { major: 0, minor: 3 }) => Ok(Header::V3(
+                        seq.next_element()?
+                            .ok_or_else(|| de::Error::missing_field("fields"))?,
+                    )),
+                    _ => Err(serde::de::Error::custom("invalid version")),
+                }
+            }
+
+            fn visit_map<V>(self, mut map: V) -> Result<Header, V::Error>
+            where
+                V: MapAccess<'de>,
+            {
+                // insert all the fields in the serde_map as the map may have out of order fields.
+                let mut serde_map: Map<String, Value> = Map::new();
+
+                while let Some(key) = map.next_key::<StructFields>()? {
+                    serde_map.insert(key.to_string(), map.next_value()?);
+                }
+
+                if let Some(v) = serde_map.get("version") {
+                    let fields = serde_map
+                        .get(&StructFields::Fields.to_string())
+                        .ok_or_else(|| de::Error::missing_field("fields"))?;
+
+                    let version =
+                        serde_json::from_value::<ResolvableChainConfigOrVersion>(v.clone())
+                            .map_err(de::Error::custom)?
+                            .chain_config;
+                    let result = match version {
+                        EitherOrVersion::Version(Version { major: 0, minor: 2 }) => Ok(Header::V2(
+                            serde_json::from_value(fields.clone()).map_err(de::Error::custom)?,
+                        )),
+                        EitherOrVersion::Version(Version { major: 0, minor: 3 }) => Ok(Header::V3(
+                            serde_json::from_value(fields.clone()).map_err(de::Error::custom)?,
+                        )),
+                        _ => Err(serde::de::Error::custom("invalid version")),
+                    };
+
+                    return result;
+                }
+
+                Ok(Header::V1(
+                    serde_json::from_value(serde_map.into()).map_err(de::Error::custom)?,
+                ))
+            }
+        }
+
+        deserializer.deserialize_struct("Header", StructFields::list(), HeaderVisitor)
+    }
+}
+
+impl Header {
+    #[allow(clippy::too_many_arguments)]
+    pub fn create(
```

Comment:
This is used in some functions where we get a version parameter. I don't know a better way to rewrite this.

Reply:
I only see it being called from tests. If that is so, can we implement it inside the test mod? It might be a little less confusing that way.

Reply:
It is also called in Header::from_info() and Header::genesis(). 
https://github.com/EspressoSystems/espresso-sequencer/pull/1637/files#diff-f28419d1e3fd809fea4b0c32d1a7ef5223c4fa94d097d31d35368b3906e9779eR786

Reply:
ok